<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="chrome=1">
  <meta name="viewport" content="width=device-width,initial-scale=1">
  <meta name="author" content="ysk24ok">
  
  <meta name="description" content="Javaの検査例外と非検査例外について自分の理解をまとめてみた。">
  <meta property="og:description" content="Javaの検査例外と非検査例外について自分の理解をまとめてみた。">
  
  <meta property="og:image" content="https://ysk24ok.github.io/assets/images/profile.jpeg">
  <meta property="og:title" content="Javaの検査例外と非検査例外について">
  <meta property="og:url" content="https://ysk24ok.github.io/2017/01/27/java_checked_exception.html">
  <meta property="og:type" content="article">
  <link href='https://fonts.googleapis.com/css?family=Chivo:900' rel='stylesheet' type='text/css'>
  <!-- MathJax -->
  <!-- http://docs.mathjax.org/en/latest/start.html#secure-access-to-the-cdn -->
  <script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      tex2jax: {
        inlineMath: [ ['$','$'], ['\\(','\\)'] ],
        processEscapes: true
      }
    });
  </script>
  <script type="text/javascript" async src="//cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
  <!-- bootstrap -->
  <link rel="stylesheet" type="text/css" href="/assets/css/bootstrap.min.css">
  <script src="/assets/js/jquery-2.1.4.min.js"></script>
  <script src="/assets/js/bootstrap.min.js"></script>
  <!-- github.io default -->
  <link rel="stylesheet" type="text/css" href="/assets/css/stylesheet.css" media="screen">
  <link rel="stylesheet" type="text/css" href="/assets/css/github-dark.css" media="screen">
  <link rel="stylesheet" type="text/css" href="/assets/css/print.css" media="print">
  <!-- syntax highlight css -->
  <link rel="stylesheet" type="text/css" href="/assets/css/syntax.css" media="screen">
  <!-- My customized css -->
  <link rel="stylesheet" type="text/css" href="/assets/css/linkpreview.css" media="screen">
  <link rel="stylesheet" type="text/css" href="/assets/css/main.css" media="screen">
  <!-- RSS -->
  <link href="/feed.xml" type="application/rss+xml" rel="alternate" title="ysk24ok.github.io" />
  <!-- Global site tag (gtag.js) - Google Analytics -->
  <script async src="https://www.googletagmanager.com/gtag/js?id=UA-131519041-1"></script>
  <script>
    window.dataLayer = window.dataLayer || [];
    function gtag(){dataLayer.push(arguments);}
    gtag('js', new Date());

    gtag('config', 'UA-131519041-1');
  </script>
  <!--[if lt IE 9]>
  <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
  <![endif]-->
  <title>Javaの検査例外と非検査例外について - ysk24ok.github.io</title>
</head>

  <body>
    <div id="container">
      <div class="inner">
        <header>
  <h1><a href="/">ysk24ok.github.io</a></h1>
</header>
<nav class="navbar navbar-default" role="navigation">
  <ul id="main-navigation" class="nav navbar-nav">
    <li><a href="/">Home</a></li>
    <li><a href="/about">About</a></li>
  </ul>
</nav>

        <div class="post-item">
          <div class="post-header">
  <h1>
    <a href="/2017/01/27/java_checked_exception.html">Javaの検査例外と非検査例外について</a>
  </h1>
  <ul class="post-tags">
    
      <li><a href="#">java</a></li>
    
      <li><a href="#">exception</a></li>
    
  </ul>
  
  <div class="post-date">posted on 27 Jan 2017</div>
  
  <hr>
</div>

          <div class="post-content">
            <ul class="social-buttons">
  <!-- hatena bookmark -->
  <li>
    <a href="http://b.hatena.ne.jp/entry/ysk24ok.github.io/2017/01/27/java_checked_exception.html" class="hatena-bookmark-button" data-hatena-bookmark-title="Javaの検査例外と非検査例外について - ysk24ok.github.io" data-hatena-bookmark-layout="simple-balloon" title="このエントリーをはてなブックマークに追加">
      <img src="https://b.st-hatena.com/images/entry-button/button-only@2x.png" alt="このエントリーをはてなブックマークに追加" width="20" height="20" style="border: none;" />
    </a>
    <script type="text/javascript" src="https://b.st-hatena.com/js/bookmark_button.js" charset="utf-8" async="async"></script>
  </li>
  <!-- twitter -->
  <li>
    <a href="https://twitter.com/share" class="twitter-share-button">Tweet</a>
    <script>!function(d,s,id){var js,fjs=d.getElementsByTagName(s)[0],p=/^http:/.test(d.location)?'http':'https';if(!d.getElementById(id)){js=d.createElement(s);js.id=id;js.src=p+'://platform.twitter.com/widgets.js';fjs.parentNode.insertBefore(js,fjs);}}(document, 'script', 'twitter-wjs');</script>
  </li>
</ul>

            <p>Javaの検査例外と非検査例外について自分の理解をまとめてみた。<br />
なおJava初心者が書いた記事であり間違っている可能性も大いにあるのでご容赦を。</p>

<!-- more -->

<h2 id="検査例外と非検査例外の違い">検査例外と非検査例外の違い</h2>

<ul>
  <li>非検査例外(unchecked exception)
    <ul>
      <li>例: NullPointerExceptionなど</li>
      <li>大体の言語には非検査例外しかない</li>
      <li>java.lang.RuntimeExceptionのサブクラス</li>
      <li>プログラミングエラーで関数が想定外の使われ方をしている場合にthrowされる例外
        <ul>
          <li>例: 型の違う変数を引数に渡している</li>
        </ul>
      </li>
      <li>プログラミングエラーになる原因はゴマンとあり予め想定できないので、<br />
callerに例外ハンドリングをさせず実行を止める</li>
    </ul>
  </li>
  <li>検査例外(checked exception)
    <ul>
      <li>例: IOExceptionなど</li>
      <li>java.lang.Exceptionのサブクラス</li>
      <li>プログラミングエラーがなく想定通りの使われ方をしてもthrowされる例外
        <ul>
          <li>例: ディスク容量に空きがないのでファイルの書き込みができない</li>
        </ul>
      </li>
      <li>予め想定されうる例外なので、callerに例外ハンドリングを強制して対処(回復)させる</li>
    </ul>
  </li>
</ul>

<h2 id="例">例</h2>

<h3 id="検査例外">検査例外</h3>

<div class="language-java highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kd">class</span> <span class="nc">Example</span> <span class="o">{</span>

    <span class="kd">public</span> <span class="kd">static</span> <span class="kt">void</span> <span class="nf">main</span> <span class="o">(</span><span class="n">String</span><span class="o">[]</span> <span class="n">args</span><span class="o">)</span> <span class="o">{</span>
        <span class="n">Example</span><span class="o">.</span><span class="na">methodA</span><span class="o">();</span>
    <span class="o">}</span>

    <span class="kd">public</span> <span class="kd">static</span> <span class="kt">void</span> <span class="nf">methodA</span><span class="o">()</span> <span class="o">{</span>
        <span class="k">throw</span> <span class="k">new</span> <span class="nf">Exception</span><span class="o">(</span><span class="s">"throws checked exception"</span><span class="o">);</span>
    <span class="o">}</span>
<span class="o">}</span>
</code></pre></div></div>

<div class="language-sh highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>javac Example.java <span class="o">&amp;&amp;</span> java Example
Example.java:8: error: unreported exception Exception<span class="p">;</span> must be caught or declared to be thrown
        throw new Exception<span class="o">(</span><span class="s2">"throws checked exception"</span><span class="o">)</span><span class="p">;</span>
        ^
1 error
</code></pre></div></div>

<p>checked exceptionは投げるだけではコンパイルエラーになる。</p>

<div class="language-java highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kd">class</span> <span class="nc">Example</span> <span class="o">{</span>

    <span class="kd">public</span> <span class="kd">static</span> <span class="kt">void</span> <span class="nf">main</span> <span class="o">(</span><span class="n">String</span><span class="o">[]</span> <span class="n">args</span><span class="o">)</span> <span class="o">{</span>
        <span class="n">Example</span><span class="o">.</span><span class="na">methodA</span><span class="o">();</span>
    <span class="o">}</span>

    <span class="kd">public</span> <span class="kd">static</span> <span class="kt">void</span> <span class="nf">methodA</span><span class="o">()</span> <span class="kd">throws</span> <span class="n">Exception</span> <span class="o">{</span>
        <span class="k">throw</span> <span class="k">new</span> <span class="nf">Exception</span><span class="o">(</span><span class="s">"throws checked exception"</span><span class="o">);</span>
    <span class="o">}</span>
<span class="o">}</span>
</code></pre></div></div>

<div class="language-sh highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>javac Example.java <span class="o">&amp;&amp;</span> java Example
Example.java:4: error: unreported exception Exception<span class="p">;</span> must be caught or declared to be thrown
        Example.methodA<span class="o">()</span><span class="p">;</span>
                       ^
1 error
</code></pre></div></div>

<p>calleeのメソッド定義に<code class="highlighter-rouge">throws</code>を書くだけではコンパイルエラー。</p>

<div class="language-java highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kd">class</span> <span class="nc">Example</span> <span class="o">{</span>

    <span class="kd">public</span> <span class="kd">static</span> <span class="kt">void</span> <span class="nf">main</span> <span class="o">(</span><span class="n">String</span><span class="o">[]</span> <span class="n">args</span><span class="o">)</span> <span class="o">{</span>
        <span class="k">try</span> <span class="o">{</span>
            <span class="n">Example</span><span class="o">.</span><span class="na">methodA</span><span class="o">();</span>
        <span class="o">}</span> <span class="k">catch</span> <span class="o">(</span><span class="n">Exception</span> <span class="n">e</span><span class="o">)</span> <span class="o">{</span>
            <span class="n">System</span><span class="o">.</span><span class="na">err</span><span class="o">.</span><span class="na">println</span><span class="o">(</span><span class="n">e</span><span class="o">.</span><span class="na">getMessage</span><span class="o">());</span>
        <span class="o">}</span>
        <span class="n">System</span><span class="o">.</span><span class="na">out</span><span class="o">.</span><span class="na">println</span><span class="o">(</span><span class="s">"finished"</span><span class="o">);</span>
    <span class="o">}</span>

    <span class="kd">public</span> <span class="kd">static</span> <span class="kt">void</span> <span class="nf">methodA</span><span class="o">()</span> <span class="o">{</span>
        <span class="k">throw</span> <span class="k">new</span> <span class="nf">Exception</span><span class="o">(</span><span class="s">"throws checked exception"</span><span class="o">);</span>
    <span class="o">}</span>
<span class="o">}</span>
</code></pre></div></div>

<div class="language-sh highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>javac Example.java <span class="o">&amp;&amp;</span> java Example
Example.java:13: error: unreported exception Exception<span class="p">;</span> must be caught or declared to be thrown
        throw new Exception<span class="o">(</span><span class="s2">"throws checked exception"</span><span class="o">)</span><span class="p">;</span>
        ^
1 error
</code></pre></div></div>

<p>callerで<code class="highlighter-rouge">catch</code>するだけでもコンパイルエラー。</p>

<div class="language-java highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kd">class</span> <span class="nc">Example</span> <span class="o">{</span>

    <span class="kd">public</span> <span class="kd">static</span> <span class="kt">void</span> <span class="nf">main</span> <span class="o">(</span><span class="n">String</span><span class="o">[]</span> <span class="n">args</span><span class="o">)</span> <span class="o">{</span>
        <span class="k">try</span> <span class="o">{</span>
            <span class="n">Example</span><span class="o">.</span><span class="na">methodA</span><span class="o">();</span>
        <span class="o">}</span> <span class="k">catch</span> <span class="o">(</span><span class="n">Exception</span> <span class="n">e</span><span class="o">)</span> <span class="o">{</span>
            <span class="n">System</span><span class="o">.</span><span class="na">err</span><span class="o">.</span><span class="na">println</span><span class="o">(</span><span class="n">e</span><span class="o">.</span><span class="na">getMessage</span><span class="o">());</span>
        <span class="o">}</span>
        <span class="n">System</span><span class="o">.</span><span class="na">out</span><span class="o">.</span><span class="na">println</span><span class="o">(</span><span class="s">"finished"</span><span class="o">);</span>
    <span class="o">}</span>

    <span class="kd">public</span> <span class="kd">static</span> <span class="kt">void</span> <span class="nf">methodA</span><span class="o">()</span> <span class="kd">throws</span> <span class="n">Exception</span> <span class="o">{</span>
        <span class="k">throw</span> <span class="k">new</span> <span class="nf">Exception</span><span class="o">(</span><span class="s">"throws checked exception"</span><span class="o">);</span>
    <span class="o">}</span>
<span class="o">}</span>
</code></pre></div></div>

<div class="language-sh highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>javac Example.java <span class="o">&amp;&amp;</span> java Example
throws checked exception
finished
</code></pre></div></div>

<p>calleeのメソッド定義に<code class="highlighter-rouge">throws</code>を加え、callerで検査例外をcatchする必要がある。</p>

<div class="language-java highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kd">class</span> <span class="nc">Example</span> <span class="o">{</span>

    <span class="kd">public</span> <span class="kd">static</span> <span class="kt">void</span> <span class="nf">main</span> <span class="o">(</span><span class="n">String</span><span class="o">[]</span> <span class="n">args</span><span class="o">)</span> <span class="o">{</span>
        <span class="k">try</span> <span class="o">{</span>
            <span class="n">Example</span><span class="o">.</span><span class="na">methodA</span><span class="o">();</span>
        <span class="o">}</span> <span class="k">catch</span> <span class="o">(</span><span class="n">Exception</span> <span class="n">e</span><span class="o">)</span> <span class="o">{</span>
            <span class="n">System</span><span class="o">.</span><span class="na">err</span><span class="o">.</span><span class="na">println</span><span class="o">(</span><span class="n">e</span><span class="o">.</span><span class="na">getMessage</span><span class="o">());</span>
        <span class="o">}</span>
        <span class="n">System</span><span class="o">.</span><span class="na">out</span><span class="o">.</span><span class="na">println</span><span class="o">(</span><span class="s">"finished"</span><span class="o">);</span>
    <span class="o">}</span>

    <span class="kd">public</span> <span class="kd">static</span> <span class="kt">void</span> <span class="nf">methodB</span><span class="o">()</span> <span class="kd">throws</span> <span class="n">Exception</span> <span class="o">{</span>
        <span class="k">throw</span> <span class="k">new</span> <span class="nf">Exception</span><span class="o">(</span><span class="s">"throws checked exception"</span><span class="o">);</span>
    <span class="o">}</span>

    <span class="kd">public</span> <span class="kd">static</span> <span class="kt">void</span> <span class="nf">methodA</span><span class="o">()</span> <span class="kd">throws</span> <span class="n">Exception</span> <span class="o">{</span>
        <span class="n">methodB</span><span class="o">();</span>
    <span class="o">}</span>
<span class="o">}</span>
</code></pre></div></div>

<p>callerでcatchせず<code class="highlighter-rouge">throws</code>を定義することで、さらに上位のcallerに例外を流すこともできる。</p>

<h3 id="非検査例外">非検査例外</h3>

<div class="language-java highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kd">class</span> <span class="nc">Example</span> <span class="o">{</span>

    <span class="kd">public</span> <span class="kd">static</span> <span class="kt">void</span> <span class="nf">main</span> <span class="o">(</span><span class="n">String</span><span class="o">[]</span> <span class="n">args</span><span class="o">)</span> <span class="o">{</span>
        <span class="n">Example</span><span class="o">.</span><span class="na">methodA</span><span class="o">();</span>
    <span class="o">}</span>

    <span class="kd">public</span> <span class="kd">static</span> <span class="kt">void</span> <span class="nf">methodA</span><span class="o">()</span> <span class="o">{</span>
        <span class="k">throw</span> <span class="k">new</span> <span class="nf">RuntimeException</span><span class="o">(</span><span class="s">"throws unchecked exception"</span><span class="o">);</span>
    <span class="o">}</span>
<span class="o">}</span>
</code></pre></div></div>

<div class="language-sh highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>javac Example.java <span class="o">&amp;&amp;</span> java Example
Exception <span class="k">in </span>thread <span class="s2">"main"</span> java.lang.RuntimeException: throws unchecked exception
        at Example.methodA<span class="o">(</span>Example.java:8<span class="o">)</span>
        at Example.main<span class="o">(</span>Example.java:4<span class="o">)</span>
</code></pre></div></div>

<p>非検査例外はcalleeで投げるだけでよく、callerはcatchしない（するべきではない？）。</p>

<h2 id="感想">感想</h2>

<p>検査例外によってcalleeの内部に閉じておくべきものが<br />
callerに見えてしまっている気がする。<br />
callerでは下位のレイヤーを意識したくないのに、対処を強制されている感じ。<br />
callerは下位レイヤーの知識がないので<br />
catchしてもどうそれに対処すればいいのかとなるような。。。</p>

<p>結局callerに例外ハンドリングを強制するのではなく、<br />
calleeで非検査例外を投げて実行を止める方がいいのでは、と思ってしまう。</p>

<h2 id="参考url">参考URL</h2>

<ul>
  <li><a href="http://ebc-2in2crc.hatenablog.jp/entry/20120729/1343557350">Java のチェック例外と非チェック例外の考察まとめ - 全力で怠けたい</a></li>
  <li><a href="http://qiita.com/yuba/items/d41290eca726559cd743">Javaの検査例外は、呼び出し側の責任でない異常系 - Qiita</a></li>
  <li><a href="http://hjm333.hatenablog.com/entry/2016/07/26/223044">【Effective Java】項目５８：回復可能な状態にはチェックされる例外を、プログラミングエラーには実行時例外を使用する - The King’s Museum</a></li>
  <li><a href="http://qiita.com/yuya_presto/items/3b651d6b0cf38f77e933">Java/Androidにおける例外設計、あるいは「契約による設計」によるシンプルさの追求 - Qiita</a></li>
</ul>

            <ul class="social-buttons">
  <!-- hatena bookmark -->
  <li>
    <a href="http://b.hatena.ne.jp/entry/ysk24ok.github.io/2017/01/27/java_checked_exception.html" class="hatena-bookmark-button" data-hatena-bookmark-title="Javaの検査例外と非検査例外について - ysk24ok.github.io" data-hatena-bookmark-layout="simple-balloon" title="このエントリーをはてなブックマークに追加">
      <img src="https://b.st-hatena.com/images/entry-button/button-only@2x.png" alt="このエントリーをはてなブックマークに追加" width="20" height="20" style="border: none;" />
    </a>
    <script type="text/javascript" src="https://b.st-hatena.com/js/bookmark_button.js" charset="utf-8" async="async"></script>
  </li>
  <!-- twitter -->
  <li>
    <a href="https://twitter.com/share" class="twitter-share-button">Tweet</a>
    <script>!function(d,s,id){var js,fjs=d.getElementsByTagName(s)[0],p=/^http:/.test(d.location)?'http':'https';if(!d.getElementById(id)){js=d.createElement(s);js.id=id;js.src=p+'://platform.twitter.com/widgets.js';fjs.parentNode.insertBefore(js,fjs);}}(document, 'script', 'twitter-wjs');</script>
  </li>
</ul>

          </div>
        </div>
        <footer>
  This page was generated by <a href="https://pages.github.com">GitHub Pages</a>. Tactile theme by <a href="https://twitter.com/jasonlong">Jason Long</a>.
</footer>

      </div>
    </div>
  </body>
</html>
